File: src/DiagToolsValidator/DiagToolsValidator.Core/Functionality/Debugging.fs
namespace DiagToolsValidator.Core.Functionality

open System
open System.IO
open System.Collections.Generic

module Debugging =
    let GenerateDebugScript (sosCommandList: string list) (scriptPath: string) =
        let debugCommandList = 
            if DotNet.CurrentRID.Contains("win")
            then
                let userProfile = Environment.GetEnvironmentVariable("USERPROFILE")
                let sosPluginPath = Path.Combine(userProfile, ".dotnet", "sos", "sos.dll")
                let preRunCommandList =
                    [
                        ".unload sos";
                        $".load {sosPluginPath}";
                    ]

                let winSOSCommandList = sosCommandList |> List.map (fun command -> $"!{command}")
                    
                let exitCommandList =
                    [
                        ".detach";
                        "qq";
                    ]
                     
                [preRunCommandList; winSOSCommandList; exitCommandList]
                |> List.concat

            else
                let preRunCommandList = List.Empty
                let exitCommandList =
                    [
                        "exit";
                    ]
                [preRunCommandList; sosCommandList; exitCommandList]
                |> List.concat
        
        try
            File.WriteAllLines(scriptPath, debugCommandList)
            Choice1Of2 scriptPath
        with ex -> Choice2Of2 ex


    let DebugDumpWithSOS (debugger: string) 
                         (env: Dictionary<string, string>)
                         (workingDirectory: string)
                         (debuggerScriptPath: string)
                         (dumpPath: string) =
        let arguments = 
            if DotNet.CurrentRID.Contains("win")
            then $"-cf {debuggerScriptPath} -z {dumpPath}"
            else $"-c {dumpPath} -s {debuggerScriptPath}"

        CommandLineTool.RunCommand debugger 
                                   arguments 
                                   workingDirectory 
                                   env 
                                   true 
                                   CommandLineTool.IgnoreOutputData
                                   CommandLineTool.IgnoreErrorData
                                   true


    let DebugAttachedProcessWithSOS (debugger: string) 
                                    (env: Dictionary<string, string>)
                                    (debuggerScriptPath: string)
                                    (pid: string) =
        let arguments = 
            if DotNet.CurrentRID.Contains("win")
            then $"-cf {debuggerScriptPath} -p {pid}"
            else $"-s {debuggerScriptPath} -p {pid}"

        CommandLineTool.RunCommand debugger 
                                   arguments 
                                   "" 
                                   env 
                                   true 
                                   CommandLineTool.IgnoreOutputData 
                                   CommandLineTool.IgnoreErrorData 
                                   true
                                   
    let DebugLaunchableWithSOS (debugger: string) 
                               (env: Dictionary<string, string>)
                               (workingDirectory: string)
                               (debuggerScriptPath: string)
                               (launchable: string) =
        let arguments = 
            if DotNet.CurrentRID.Contains("win")
            then $"-g -cf {debuggerScriptPath} {launchable}"
            else $"lldb -s {debuggerScriptPath} -o \"run\" {launchable}"

        let _env = new Dictionary<string, string>(env)
        _env["DOTNET_StressLogLevel"] <- "10"
        _env["DOTNET_TotalStressLogSize"] <- "8196"
        CommandLineTool.RunCommand debugger 
                                   arguments 
                                   workingDirectory 
                                   _env 
                                   true 
                                   CommandLineTool.IgnoreOutputData
                                   CommandLineTool.IgnoreErrorData
                                   true

File: src/DiagToolsValidationRunner/DiagToolsValidationRunner.Core/Functionality/Debugging.cs
﻿namespace DiagToolsValidationRunner.Core.Functionality
{
    public class CLIDebugger
    {
        public string CLIDebuggerPath { get; }
        
        public CLIDebugger(string cliDebuggerPath)
        {
            this.CLIDebuggerPath = cliDebuggerPath;
        }

        public static void GenerateSOSDebuggingScript(string debuggingScriptPath, List<string> SOSDebugCommandList)
        {
            // Generate debug script
            List<string> preRunCommandList = new();
            List<string> sosCommandList = new();
            List<string> exitCommandList = new();
            if (OperatingSystem.IsWindows())
            {
                string? userProfile = Environment.GetEnvironmentVariable("USERPROFILE");
                string sosExtension = Path.Combine($"{userProfile}", ".dotnet", "sos", "sos.dll");
                preRunCommandList = [
                    ".unload sos",
                    $".load {sosExtension}"
                ];
                sosCommandList = SOSDebugCommandList
                    .Select(command => $"!{command}")
                    .ToList();
                exitCommandList = [
                    ".detach",
                    "qq"
                ];
            }
            else
            {
                sosCommandList = SOSDebugCommandList;
                exitCommandList = ["exit"];
            }

            List<string> debuggingCommandList = preRunCommandList
                .Concat(sosCommandList)
                .Concat(exitCommandList)
                .ToList();

            File.WriteAllLines(debuggingScriptPath, debuggingCommandList);
        }

        public CommandInvokeResult DebugDump(Dictionary<string, string> env,
                                             string targetRID,
                                             string workingDirectory,
                                             string dumpPath,
                                             string debuggerScriptPath,
                                             bool redirectStdOutErr = true,
                                             bool silent = true)
        {
            string arguments =
                targetRID.Contains("win") switch
                {
                    true => $"-cf {debuggerScriptPath} -z {dumpPath}",
                    false => $"-c {dumpPath} -s {debuggerScriptPath} --batch"
                };

            CommandInvoker invoker = new(CLIDebuggerPath,
                                         arguments,
                                         env,
                                         workingDirectory,
                                         redirectStdOutErr,
                                         silent);
            return invoker.WaitForResult();
        }

        public CommandInvokeResult DebugAttachedProcess(Dictionary<string, string> env,
                                                        string targetRID,
                                                        string workingDirectory,
                                                        int pid,
                                                        string debuggerScriptPath,
                                                        bool redirectStdOutErr = true,
                                                        bool silent = true)
        {
            string arguments =
                targetRID.Contains("win") switch
                {
                    true => $"-cf {debuggerScriptPath} -p {pid}",
                    false => $"-s {debuggerScriptPath} -p {pid}  --batch"
                };

            CommandInvoker invoker = new(CLIDebuggerPath,
                                         arguments,
                                         env,
                                         workingDirectory,
                                         redirectStdOutErr,
                                         silent);
            return invoker.WaitForResult();
        }

        public CommandInvokeResult DebugLaunchable(Dictionary<string, string> env,
                                                   string targetRID,
                                                   string workingDirectory,
                                                   string launchable,
                                                   string debuggerScriptPath,
                                                   bool redirectStdOutErr = true,
                                                   bool silent = true)
        {
            string arguments =
                targetRID.Contains("win") switch
                {
                    true => $"-g -cf {debuggerScriptPath} {launchable}",
                    false => $"-s {debuggerScriptPath} --batch {launchable}"
                };

            CommandInvoker invoker = new(CLIDebuggerPath,
                                         arguments,
                                         env,
                                         workingDirectory,
                                         redirectStdOutErr,
                                         silent);
            return invoker.WaitForResult();
        }
    }

    public class DotNetDumpAnalyzer
    {
        public string DotNetDumpILPath { get; }

        public DotNetDumpAnalyzer(string dotNetDumpILPath)
        {
            this.DotNetDumpILPath = dotNetDumpILPath;
        }

        public CommandInvokeResult DebugDump(Dictionary<string, string> env,
                                             string workingDirectory,
                                             string dumpPath,
                                             List<string> sosCommandList,
                                             bool redirectStdOutErr = true,
                                             bool silent = true)
        {
            string dotNetExecutablePath = DotNetInfrastructure.GetDotNetExecutableFromEnv(env);
            CommandInvoker invoker = new(dotNetExecutablePath,
                                         $"{DotNetDumpILPath} analyze {dumpPath}",
                                         env,
                                         workingDirectory,
                                         redirectStdOutErr,
                                         silent);
            foreach (var command in sosCommandList)
            {
                invoker.StandardInput.WriteLine(command);
            }

            invoker.StandardInput.WriteLine("exit");
            return invoker.WaitForResult();
        }
    }
}
